File: simple_rag.ts
/**
 * Concise RAG pipeline in TypeScript (LangChain + Hugging Face embeddings)
 * - Reads 3 local Word docs
 * - Chunks text
 * - Embeds using HuggingFaceEmbeddings
 * - Stores in FAISS (in-memory) or Chroma (persistent)
 * - Runs a sample retrieval query using OpenAI LLM
 */

import * as fs from "fs";
import * as path from "path";
import "dotenv/config";

import { Document } from "@langchain/core/documents";
import { RecursiveCharacterTextSplitter } from "@langchain/textsplitters";
import { HuggingFaceTransformersEmbeddings } from "@langchain/community/embeddings/huggingface";
import { OpenAI } from "@langchain/openai";
import { RetrievalQAChain } from "langchain/chains";
import { Chroma } from "@langchain/community/vectorstores/chroma";
import { FaissStore } from "@langchain/community/vectorstores/faiss";
import { load } from "docx-parser";

const FILES = ["doc1.docx", "doc2.docx", "doc3.docx"]; // Local docx files
const VECTOR_STORE: "faiss" | "chroma" = "faiss"; // change to "chroma" if needed
const CHROMA_DIR = "./chroma_store";
const EMBEDDING_MODEL = "Xenova/all-MiniLM-L6-v2"; // Hugging Face model
const CHUNK_SIZE = 1000;
const CHUNK_OVERLAP = 200;

(async () => {
  console.log("Starting RAG pipeline...");

  // --- 1. Read documents ---
  const docs: Document[] = [];
  for (const file of FILES) {
    const filePath = path.resolve(file);
    const text = await load(filePath);
    docs.push(new Document({ pageContent: text }));
  }

  // --- 2. Chunk documents ---
  const splitter = new RecursiveCharacterTextSplitter({
    chunkSize: CHUNK_SIZE,
    chunkOverlap: CHUNK_OVERLAP,
  });
  const chunks = await splitter.splitDocuments(docs);
  console.log(`Total chunks: ${chunks.length}`);

  // --- 3. Embeddings ---
  const embeddings = new HuggingFaceTransformersEmbeddings({
    modelName: EMBEDDING_MODEL,
  });

  // --- 4. Vector store ---
  let vectorStore;
  if (VECTOR_STORE === "faiss") {
    console.log("Using FAISS (in-memory) vector store...");
    vectorStore = await FaissStore.fromDocuments(chunks, embeddings);
  } else {
    console.log("Using Chroma (persistent) vector store...");
    vectorStore = await Chroma.fromDocuments(chunks, embeddings, {
      collectionName: "rag_docs",
      url: undefined,
      collectionMetadata: {},
      persistDirectory: CHROMA_DIR,
    });
  }

  // --- 5. LLM (OpenAI) ---
  const llm = new OpenAI({
    modelName: "gpt-4o-mini",
    temperature: 0,
  });

  // --- 6. Create retrieval chain ---
  const retriever = vectorStore.asRetriever({ k: 4 });
  const chain = RetrievalQAChain.fromLLM(llm, retriever);

  // --- 7. Example query ---
  const query = "Summarize the main ideas from these documents.";
  console.log("\nQuery:", query);
  const result = await chain.call({ query });
  console.log("\nAnswer:\n", result.text);

  console.log("\n✅ RAG pipeline complete.");
})();

File: simple_rag.py
"""
Concise RAG pipeline using LangChain + HuggingFace embeddings.
Reads 3 local Word docs, chunks & embeds them, stores vectors in FAISS (in-memory)
or Chroma (persistent), then runs a sample query using RetrievalQA.
"""

import os
from pathlib import Path
import docx

from langchain.docstore.document import Document
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.vectorstores import FAISS, Chroma
from langchain.chains import RetrievalQA
from langchain.llms import OpenAI  # example LLM (swap if needed)

# --- Config ---
FILES = ["doc1.docx", "doc2.docx", "doc3.docx"]  # local Word docs
VECTOR_STORE = "faiss"  # choose "faiss" or "chroma"
EMBEDDING_MODEL = "sentence-transformers/all-MiniLM-L6-v2"
CHROMA_PERSIST_DIR = "./chroma_store"
CHUNK_SIZE = 1000
CHUNK_OVERLAP = 200

# --- 1. Load documents ---
docs = []
for file in FILES:
    text = "\n\n".join(p.text for p in docx.Document(file).paragraphs if p.text.strip())
    docs.append(Document(page_content=text))

# --- 2. Chunk documents ---
splitter = RecursiveCharacterTextSplitter(chunk_size=CHUNK_SIZE, chunk_overlap=CHUNK_OVERLAP)
chunks = []
for d in docs:
    for chunk in splitter.split_text(d.page_content):
        chunks.append(Document(page_content=chunk))
print(f"Total chunks: {len(chunks)}")

# --- 3. Embeddings ---
embeddings = HuggingFaceEmbeddings(model_name=EMBEDDING_MODEL)

# --- 4. Vector store ---
if VECTOR_STORE == "faiss":
    print("Using in-memory FAISS store...")
    vs = FAISS.from_documents(chunks, embeddings)
elif VECTOR_STORE == "chroma":
    print(f"Using persistent Chroma store at {CHROMA_PERSIST_DIR} ...")
    os.makedirs(CHROMA_PERSIST_DIR, exist_ok=True)
    vs = Chroma.from_documents(chunks, embeddings, persist_directory=CHROMA_PERSIST_DIR)
    vs.persist()
else:
    raise ValueError("VECTOR_STORE must be 'faiss' or 'chroma'")

# --- 5. RetrievalQA ---
llm = OpenAI(temperature=0)
qa = RetrievalQA.from_chain_type(llm=llm, retriever=vs.as_retriever(search_kwargs={"k": 4}))

# --- 6. Example query ---
query = "Summarize the main ideas from these documents."
print("\nQuery:", query)
answer = qa.run(query)
print("\nAnswer:\n", answer)
